### PR TITLE
feat: Add fromDOM method

### PR DESCRIPTION
**Pull request by doubleface**
To get a CozyClient instance from data injected by the stack in the DOM
in the main page

fixes issue #799

### CODE REVIEW

File: packages/cozy-client/src/CozyClient.spec.js [hunk 1]
```diff
@@ -153,6 +153,15 @@ describe('CozyClient initialization', () => {
     expect(client.stackClient.token.token).toBe(token)
   })
 
+  it('can be instantiated from dataset injected by the Stack in DOM', async () => {
+      const options = { cozyDomain: 'cozy.tools', cozyToken: 'abc123' }
+      const globalQuerySelectorBefore = document.querySelector
+      document.querySelector = jest.fn().mockReturnValue({ dataset: options })
```
Comment (ptbrowne) [inline packages/cozy-client/src/CozyClient.spec.js:159]: I think you want to use jest.spyOn to temporarily replace document.querySelector.

```
beforeEach(() => {
  jest.spyOn(document, 'querySelector').mockReturnValue({ dataset: options })
})

afterEach(() => {
  document.querySelector.mockRestore()
})
```

File: packages/cozy-client/src/CozyClient.js [hunk 2]
```diff
@@ -239,6 +239,36 @@ class CozyClient {
...
+   */
+  static fromDOM(selector = '[role=application]', options = {}) {
+    const root = document.querySelector(selector)
+    if (!root || !root.dataset) {
+      throw new Error(
+        `Found no data in ${selector} to instantiate cozyClient`
+      )
+    }
+
+    const {cozyDomain, cozyToken} = root.dataset
```
Comment (Crash--) [inline packages/cozy-client/src/CozyClient.js:257]: We should supports the both way the stack injects the content : 
- in the dataset 
- but also in JSON format in data-cozy 

```suggestion
const data = root.dataset.cozy
      ? JSON.parse(root.dataset.cozy)
      : { ...root.dataset }
const {cozyDomain, cozyToken} = data
```

File: docs/api/cozy-client.md [hunk 3]
```diff
@@ -1147,6 +1148,19 @@ Warning: unlike other instantiators, this one needs to be awaited.
 In konnector/service context, CozyClient can be instantiated from environment variables
 
 **Kind**: static method of [<code>CozyClient</code>](#CozyClient)  
+<a name="CozyClient.fromDOM"></a>
+
+### CozyClient.fromDOM(selector, options) ⇒ <code>object</code>
+In cozy app context, CozyClient can be instantiated from data injected by the stack
```
Comment (ptbrowne) [inline docs/api/cozy-client.md:1154]:
```suggestion
When used from an app, CozyClient can be instantiated from the data injected by the stack in
the DOM.
```

Comment (Crash--) [reply docs/api/cozy-client.md]: When used from an web app ? ;)

File: packages/cozy-client/src/CozyClient.js [hunk 4]
```diff
@@ -239,6 +239,38 @@ class CozyClient {
...
   }
 
+  /**
+   * In cozy app context, CozyClient can be instantiated from data injected by the stack
+   *
+   * @param  {string}   selector - Options
+   * @param  {object}   options  - CozyClient constructor options
+   * @returns {object} - CozyClient instance
+   */
+  static fromDOM(selector = '[role=application]', options = {}) {
```
Comment (ptbrowne) [inline packages/cozy-client/src/CozyClient.js:249]: I am hesitant on the signature : I think selector will not be used as frequently as the options imho. The problem is that to define options, signature has to be passed, it is a bit inconvenient. I wonder if we should exchange the two args ? It the same problem for fromEnv though.

File: packages/cozy-client/src/CozyClient.js [hunk 5]
```diff
@@ -239,6 +239,38 @@ class CozyClient {
     })
   }
 
+  /**
+   * In cozy app context, CozyClient can be instantiated from data injected by the stack
```
Comment (ptbrowne) [inline packages/cozy-client/src/CozyClient.js:243]:
```suggestion
   * When used from an app, CozyClient can be instantiated from the data injected by the stack    in
   * the DOM.
```

**Review by ptbrowne — APPROVED** [review]

File: packages/cozy-client/src/CozyClient.js (hunk 4 above)
Comment (doubleface) [reply packages/cozy-client/src/CozyClient.js:249]: I agree, but I did it this way for consistency. 
I don't have any use case for the option parameter though.

Comment (ptbrowne) [reply packages/cozy-client/src/CozyClient.js:250]: Passing the schema, the oauth options, the app name.

**Pull request merged**